### ISSUE TITLE
Add buttons to UI

### ISSUE THREAD

**Issue by torkleyy** [amethyst/amethyst#577]
Implementing `UiButton` sounds like a good first issue. I'm opening this issue for discussion how to implement it.

@Xaeroxe Maybe you could mentor this for @Boiethios as they have interest in fixing this issue.

**Comment by Xaeroxe** [comment]
Partially fixed in #580 

Still needs a proper `UiButton` structure though.

**Comment by jojolepro** [comment]
or macro or prefab.

**Comment by Hittherhod** [comment]
I'd be willing to tackle this with mentoring. I'm been at a loss for how to do this while trying to make a menu.

**Comment by jojolepro** [comment]
I can help you with that, hit me up on gitter.

**Comment by jojolepro** [comment]
@Hittherhod So, basically I was going to wait to have a layout system to do it, but its not really hard to make without.
What you have to do is to create a text over a fixed image. To make the text be over the image, you can simply change the z. Optionally, you can set the text be a child of the button image. The other thing you have to do is center the text in the middle of the image (x,y). To make the button image clickable, see #580 in the ui example. The ui event is Clicked in this case.

When you can do that, you can either make a macro to create the entity, or use a struct with the builder pattern.

**Comment by azriel91** [comment]
To centre the text, I'm guessing you'd have to ask the font to calculate the width it needs, given the string + font size?

More questions (more so to satisfy my curiosity):

* Should this also cover the "what happens when the user hovers over the button"? Do we say "provide us with what you want to do" such as change the background image?
* How about using a single fill colour instead of an image?

**Comment by jojolepro** [comment]
>Should this also cover the "what happens when the user hovers over the button"? Do we say "provide us with what you want to do" such as change the background image?

When the user hovers: Depends on the button settings. We can make it either that the user specifies what happens, and we provide a decent default, or we just go for the classic "swap image by hovered image".

>How about using a single fill colour instead of an image?
Its up to the user to customize the button. For the default, I'm not sure if its better to have a basic button picture, or just a color fill. I guess that's up to you to decide that :)

TLDR: User chooses everything, but we provide decent default for quick iteration.

**Comment by Hittherhod** [comment]
@jojolepro Sorry about the delay; I've been without power since volunteering. For the default I was thinking of just having a black rectangle with a white fill: is there a simple way to do this with gfx-rs? Everything I find just jumps straight into multiple triangles and shading things.

**Comment by Hittherhod** [comment]
@jojolepro Also, how do I make the text a child of the button image in Rust?

**Comment by jojolepro** [comment]
base yourself on this PR https://github.com/jojolepro/amethyst/tree/uilayout

**Comment by Rhuagh** [comment]
I'd recommend some sort of Grey or beige color for the default, black and white have a tendency to disappear into other things :) Might be worthwhile having easy ways of modifying the colors from the user perspective also, even if it's for a "default" setup

**Comment by Rhuagh** [comment]
For rendering you can use 2 `UiImage`s, with pure color textures, where one is anchored to the center of the other and stretched with a margin. Check the ui layout pr

**Comment by Rhuagh** [comment]
A different solution would be a single image with a texture that looks like a button

**Comment by torkleyy** [comment]
Just to make sure we don't duplicate work: I think @Hittherhod is working on this.

**Comment by Hittherhod** [comment]
I think this can be closed as #613 is merged
